src/bootstrap/flags.rs

```diff
@@ -89,21 +104,41 @@ impl Flags {
         opts.optopt("j", "jobs", "number of jobs to run in parallel", "JOBS");
         opts.optflag("h", "help", "print this help message");
 
-        let usage = |n, opts: &Options| -> ! {
-            let command = args.get(0).map(|s| &**s);
-            let brief = format!("Usage: x.py {} [options] [<args>...]",
-                                command.unwrap_or("<command>"));
+        // fn usage()
+        let usage = |exit_code: i32, opts: &Options, subcommand_help: &str, extra_help: &str| -> ! {
+            println!("{}", opts.usage(subcommand_help));
+            if !extra_help.is_empty() {
+                println!("{}", extra_help);
+            }
+            process::exit(exit_code);
+        };
+
+        // We can't use getopt to parse the options until we have completed specifying which
+        // options are valid, but under the current implementation, some options are conditional on
+        // the subcommand. Therefore we must manually identify the subcommand first, so that we can
+        // complete the definition of the options.  Then we can use the getopt::Matches object from
+        // there on out.
+        let mut possible_subcommands = args.iter().collect::<Vec<_>>();
+        possible_subcommands.retain(|&s| !s.starts_with('-'));
+        let subcommand = match possible_subcommands.first() {
```

Comment:
Good catch!  Yet _another_ reason getopts should have some `parse()` variant that will not error out if an invalid option is encountered.  Then we could just parse once for the subcommand, add more options, and then parse again.

Hmmm...options with their own arguments.  That's actually a tough nut to crack without rewriting a lot of our own getopt-like logic.

What if we cheated and took the first argument not starting with `-` AND that is one of our valid subcommands?  Is there any overlap between arguments to options and subcommands?

I'll see what I can do to fix this later tonight.